### PR TITLE
ConvolutionOp multiply and add fusing patterns

### PR DESCRIPTION
**Pull request by mvasiljevicTT**
### Ticket
[Link to Github Issue](https://github.com/tenstorrent/tt-mlir/issues/4155)

### Problem description
CNN models used through torch-xla have `ttir.convolution` ops, so decomposed batch norms coming after them could not be fused with existing `conv2dOp` fusing patterns.

### What's changed
- Generalized existing `conv2dOp` fusing patterns to also handle `convolutionOp`
- Fixed bias handling in `ConvolutionToConv2d` pattern
- Added tests for resnet submodules verifying decomposed batchnorm ops are moved before convolution.

### Checklist
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

File: lib/Dialect/TTIR/Transforms/TTIRFusing.cpp [hunk 1]
```diff
@@ -17,71 +17,100 @@ namespace mlir::tt::ttir {
 
 namespace {
 // Check if we can fuse conv2d followed by add into conv2d with bias.
```
Comment (odjuricicTT) [inline lib/Dialect/TTIR/Transforms/TTIRFusing.cpp]: Since all occurrences of conv2d were changes to conv, this one should be as well.

**Comment by mtopalovicTT** [comment]
@mvasiljevicTT afaik when we execute ttnn pipeline this is the sequence of passes that are run `TTIRFusing` -> `TTIRToTTIRDecomposition` -> `TTIRFusing`. Shouldn't second fusing pass pickup rewriten convolution op and fuse the decomposed bn?

**Comment by mvasiljevicTT** [comment]
> @mvasiljevicTT afaik when we execute ttnn pipeline this is the sequence of passes that are run `TTIRFusing` -> `TTIRToTTIRDecomposition` -> `TTIRFusing`. Shouldn't second fusing pass pickup rewriten convolution op and fuse the decomposed bn?

@mtopalovicTT Ideally this would be the case, but along the way additional reshapes, broadcasts, and permutes are inserted, which the current `EraseInverseOps` doesn’t fully resolve. Updating that pass in a general way would be much more complex, so we chose this approach for now.

File: lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp [hunk 2]
```diff
@@ -455,6 +455,21 @@ struct ConvolutionToConv2dPattern : public ConvolutionDecompositionPattern {
         weightOutputShape, weightType.getElementType(),
         weightType.getEncoding(), weight, kernelPermutation);
 
+    // If bias is provided, it needs to be reshaped to match the expected shape
+    Value biasValue = adaptor.getBias();
+    if (biasValue) {
+      auto biasType = mlir::cast<RankedTensorType>(biasValue.getType());
+      auto biasPermutation = generateConvPermutation(op, conv2dLayout);
```
Comment (mtopalovicTT) [inline lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp:462]: Can you add verification in convolutionop::verify that checks that all dims are 1's except output feature dim?  Only that way this reshape makes sense.

File: lib/Dialect/TTIR/Transforms/TTIRFusing.cpp [hunk 3]
```diff
@@ -481,7 +535,16 @@ class Conv2dWithMultiply : public mlir::OpRewritePattern<MultiplyOp> {
     // Swap first and last dimensions.
     assert(newShape.size() == 4 &&
            "Scale tensor must have 4 dimensions for reshaping.");
-    std::swap(newShape[0], newShape[3]);
+
+    if constexpr (std::is_same_v<ConvOpType, Conv2dOp>) {
+      std::swap(newShape[0], newShape[3]);
+    } else {
```
Comment (mtopalovicTT) [inline lib/Dialect/TTIR/Transforms/TTIRFusing.cpp]: same here, else if constexpr and follow up with else branch with llvm_unreachable

File: lib/Dialect/TTIR/Transforms/TTIRFusing.cpp [hunk 4]
```diff
@@ -461,10 +513,12 @@ class Conv2dWithMultiply : public mlir::OpRewritePattern<MultiplyOp> {
   // and then we create new broadcast operation with the new reshaped scale
   // which broadcasts the reshaped scale to the shape of the weight tensor.
   static Value createReshapedScale(mlir::PatternRewriter &rewriter,
-                                   Location loc, Value scaleValue,
-                                   RankedTensorType weightType) {
+                                   Value scaleValue, ConvOpType convOp) {
```
Comment (mtopalovicTT) [inline lib/Dialect/TTIR/Transforms/TTIRFusing.cpp:534]: Update the comment in function to explain how the reshape is done in case of convolution op.
And also this comment below:
`    // Create a new shape (out_channels, 1, 1, 1) from (1, 1, 1, out_channels).`

File: lib/Dialect/TTIR/Transforms/TTIRFusing.cpp [hunk 5]
```diff
@@ -17,71 +17,100 @@ namespace mlir::tt::ttir {
...
+        return false;
+      }
+      auto outputFeatureDim =
+          convOp.getConvolutionLayoutAttr().getOutputFeatureDimension();
+      auto biasShape = bias.getType().getShape();
+      auto outputShape =
+          mlir::cast<mlir::RankedTensorType>(convOp.getOutput().getType())
+              .getShape();
+
+      if (biasShape.size() != outputShape.size() ||
```
Comment (mtopalovicTT) [inline lib/Dialect/TTIR/Transforms/TTIRFusing.cpp]: @LPanosTT do you know if we can add this check (i.e `biasShape.size() == outputShape.size()`) in verifier of `ConvoutionOp`? We already are already checking if bias is 4d in verifier so output shape should be 4d also right?

File: lib/Dialect/TTIR/Transforms/TTIRFusing.cpp [hunk 6]
```diff
@@ -17,71 +17,100 @@ namespace mlir::tt::ttir {
...
     return mlir::success();
   }
 
 private:
-  bool isFusable(ttir::Conv2dOp conv2dOp,
+  bool isFusable(ConvOpType convOp,
                  mlir::TypedValue<mlir::RankedTensorType> bias) const {
-    return conv2dOp && !conv2dOp.getBias() && conv2dOp->hasOneUse() &&
-           conv2dOp.isBiasCompatible(bias.getType().getShape());
+    if constexpr (std::is_same_v<ConvOpType, Conv2dOp>) {
```
Comment (mtopalovicTT) [inline lib/Dialect/TTIR/Transforms/TTIRFusing.cpp:70]: Something like this maybe:
```cpp
  bool isFusable(ConvOpType convOp,
                 mlir::TypedValue<mlir::RankedTensorType> bias) const {
		if (!convOp || convOp.getBias() || !convOp->hasOneUse()) {
			return false;
		}

		size_t outChannelDim = 0;
		if constexpr (std::is_same_v<ConvOpType, Conv2dOp>) {
			outChannelDim = 3;
		} else if constexpr (std::is_same_v<ConvOpType, ConvolutionOp>) {
			outChannelDim = convOp.getConvolutionLayoutAttr().getOutputFeatureDimension();
		} else {
              static_assert(ttmlir::utils::always_false<ConvOpType>(), "Unsupported ConvOpType");
        }

		int64_t channelDimSize = convOp.getResult().getType().getShape()[outChannelDim];
		for (auto [dim, dimSize] : llvm::enumerate(bias.getType().getShape())) {
			if (dim == outChannelDim && dimSize != channelDimSize) {
				return false;
			}
			if (dim != outChannelDim && dimSize	!= 1) {
				return false;
			}
		}
		
		return true;
  }

```

File: lib/Dialect/TTIR/Transforms/TTIRFusing.cpp [hunk 7]
```diff
@@ -434,20 +462,44 @@ class Conv2dWithMultiply : public mlir::OpRewritePattern<MultiplyOp> {
...
   }
 
   // Scale must have rank 4 and shape (1, 1, 1, out_channels).
-  static bool hasValidScaleShape(Conv2dOp convOp, RankedTensorType scaleType) {
-    return scaleType.getRank() == 4 && scaleType.getDimSize(0) == 1 &&
-           scaleType.getDimSize(1) == 1 && scaleType.getDimSize(2) == 1 &&
-           scaleType.getDimSize(3) == convOp.getOutputChannelSize();
+  static bool hasValidScaleShape(ConvOpType convOp,
+                                 RankedTensorType scaleType) {
+    if constexpr (std::is_same_v<ConvOpType, Conv2dOp>) {
```
Comment (mtopalovicTT) [inline lib/Dialect/TTIR/Transforms/TTIRFusing.cpp:494]: Constrains are the same its just feature dim position is different, so something like the suggestion for isFusable in bias should be possible.

File: lib/Dialect/TTIR/Transforms/TTIRFusing.cpp [hunk 8]
```diff
@@ -481,7 +535,16 @@ class Conv2dWithMultiply : public mlir::OpRewritePattern<MultiplyOp> {
...
+
+    if constexpr (std::is_same_v<ConvOpType, Conv2dOp>) {
+      std::swap(newShape[0], newShape[3]);
+    } else {
+      auto outputFeatureDim =
+          convOp.getConvolutionLayoutAttr().getOutputFeatureDimension();
+      auto kernelOutputFeatureDim =
+          convOp.getConvolutionLayoutAttr().getKernelOutputFeatureDimension();
+      std::swap(newShape[outputFeatureDim], newShape[kernelOutputFeatureDim]);
+    }
```
Comment (mtopalovicTT) [inline lib/Dialect/TTIR/Transforms/TTIRFusing.cpp:569]: nit empty line below

File: lib/Dialect/TTIR/Transforms/TTIRFusing.cpp (hunk 5 above)
Comment (LPanosTT) [reply lib/Dialect/TTIR/Transforms/TTIRFusing.cpp]: The fact that we check the bias shape to be 2d is actually a mistake and I should not have put that in the verifier. `ConvolutionOp` is general enough to represent any convolution. i.e `Conv1d, Conv2d, Conv3d, Conv4d, ....., ConvTranspose1d, ConvTranspose2d, ConvTranspose3d, ....`

The bias should only have data along one dimension since there is always one feature (channel) dimension. So realistically it doesn't matter what rank the bias tensor is as long as the shape == 1 in all dimension except one, and the size of that dimension must be equal to the output channel. We can always reshape to whatever the requirements of `conv2d` need the bias rank to be upon conversion.

I have a PR up that is sort of related to convolutions, I can remove this check there: https://github.com/tenstorrent/tt-mlir/pull/4667

Comment (LPanosTT) [reply lib/Dialect/TTIR/Transforms/TTIRFusing.cpp]: `ttir::ConvolutionOp` is meant to be a 1-1 of `stablehlo::ConvolutionOp`, with the addition of being able to have a bias, which `stablehlo::ConvolutionOp` does not. 

So if we get a convolution with bias via a stablehlo graph, then the bias is represented as an explicit add afterwards, in which case it will be 4D with the data along the proper dim. Which is why this hasn't caused any trouble yet.

File: lib/Dialect/TTIR/Transforms/TTIRFusing.cpp (hunk 6 above)
Comment (mvasiljevicTT) [reply lib/Dialect/TTIR/Transforms/TTIRFusing.cpp:70]: I like it, changed now 😄

File: lib/Dialect/TTIR/Transforms/TTIRFusing.cpp (hunk 7 above)
Comment (mvasiljevicTT) [reply lib/Dialect/TTIR/Transforms/TTIRFusing.cpp:494]: Changed too 👍

File: lib/Dialect/TTIR/Transforms/TTIRFusing.cpp (hunk 4 above)
Comment (mvasiljevicTT) [reply lib/Dialect/TTIR/Transforms/TTIRFusing.cpp:534]: Thanks, done 👍

File: lib/Dialect/TTIR/Transforms/TTIRFusing.cpp (hunk 3 above)
Comment (mvasiljevicTT) [reply lib/Dialect/TTIR/Transforms/TTIRFusing.cpp]: changed now

File: lib/Dialect/TTIR/Transforms/TTIRFusing.cpp (hunk 6 above)
Comment (mtopalovicTT) [reply lib/Dialect/TTIR/Transforms/TTIRFusing.cpp:70]: Can you replace `llvm_unreachable`  here and in below references with:
`static_assert(ttmlir::utils::always_false<ConvOpType>(), "Unsupported ConvOpType");`
`llvm_unreachable` is runtime assert so `static_assert` is better in this case.

File: lib/Dialect/TTIR/Transforms/TTIRFusing.cpp (hunk 5 above)
Comment (mvasiljevicTT) [reply lib/Dialect/TTIR/Transforms/TTIRFusing.cpp]: In sync with @LPanosTT I added check in verifier that bias has the same rank as the output and matches the output feature dimension, with size 1 in all other dims.

@mtopalovicTT in the snippet you are referring to `biasShape` is a shape of a *potential* bias and could still differ from `outputShape` in which case `is_fusable` should return false

File: lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp (hunk 2 above)
Comment (mvasiljevicTT) [reply lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp:462]: Added check in verify that bias has the same rank as the output and matches the output feature dimension, with size 1 in all other dims.

File: lib/Dialect/TTIR/Transforms/TTIRFusing.cpp (hunk 6 above)
Comment (mvasiljevicTT) [reply lib/Dialect/TTIR/Transforms/TTIRFusing.cpp:70]: thanks, updated

**Review by odjuricicTT — APPROVED** [review]
Looks good!

File: lib/Dialect/TTIR/Transforms/TTIRFusing.cpp (hunk 1 above)
Comment (mvasiljevicTT) [reply lib/Dialect/TTIR/Transforms/TTIRFusing.cpp]: fixed
